singularity-eos/eos/eos_base.hpp

```diff
@@ -836,6 +838,44 @@ class EosBase {
     PORTABLE_ALWAYS_THROW_OR_ABORT(msg);
   }
 
+  // JMM: This method is often going to be overloaded for special cases.
+  template <typename Indexer_t = Real *>
+  PORTABLE_INLINE_FUNCTION void
+  DensityEnergyFromPressureTemperature(const Real press, const Real temp,
+                                       Indexer_t &&lambda, Real &rho, Real &sie) const {
+    // TODO(JMM): A lot hardcoded in here... Hopefully relevent EOS's
+    // overwrite.
+    constexpr Real MAXFAC = 1e8;
+    constexpr Real EPS = 10 * robust::EPS();
+    constexpr Real MINR_DEFAULT = 1e-4;
+    constexpr Real DEFAULT_RHO_GUESS = 4;
+    using RootFinding1D::findRoot; // more robust but slower. Better default.
+    using RootFinding1D::Status;
+    CRTP copy = *(static_cast<CRTP const *>(this));
+    auto PofRT = [&](const Real r) {
+      return copy.PressureFromDensityTemperature(r, temp, lambda);
+    };
+    // JMM: This can't be zero, in case MinimumDensity is zero
+    Real rhomin = std::max(MINR_DEFAULT, copy.MinimumDensity());
+    Real rhomax = MAXFAC * rhomin;
+    Real rhoguess = rho; // use input density
+    if ((rhoguess < rhomin) || (rhoguess > rhomax)) {
+      if ((rhomin < DEFAULT_RHO_GUESS) && (rhomax > DEFAULT_RHO_GUESS)) {
+        rhoguess = DEFAULT_RHO_GUESS;
+      } else {
+        rhoguess = 0.5 * (rhomin + rhomax);
+      }
+    }
+    auto status = findRoot(PofRT, press, rhoguess, rhomin, rhomax, EPS, EPS, rho);
+    // JMM: This needs to not fail and instead return something sane
+    if (status != Status::SUCCESS) {
+      PORTABLE_WARN("DensityEnergyFromPressureTemperature failed to find root\n");
+      rho = rhoguess;
+    }
```

Comment:
This is a function that may be passed data and I see it as similar to a PTE solve. Even if it fails out, something reasonable must be done.